metadata/src/main/java/org/apache/kafka/controller/QuorumController.java

```diff
@@ -366,15 +367,15 @@ void createSnapshotGenerator(long committedOffset, int committedEpoch, long comm
                     )
                 );
             }
-            Optional<SnapshotWriter<ApiMessageAndVersion>> writer = raftClient.createSnapshot(
+            Optional<RecordsSnapshotWriter<ApiMessageAndVersion>> writer = raftClient.createSnapshot(
                 committedOffset,
                 committedEpoch,
                 committedTimestamp
             );
             if (writer.isPresent()) {
                 generator = new SnapshotGenerator(
                     logContext,
-                    writer.get(),
+                        (SnapshotWriter<ApiMessageAndVersion>) writer.get(),
```

Comment:
Please remove this cast. With the changes I suggested in my other comments the type of `writer` will be `Optional<SnapshotWriter>`. `SnapshotGenerator` should expect the type `SnapshotWriter<ApiMessageAndVersion>` for the second argument to the constructor.